src/traces/parcats/defaults.js

```diff
@@ -0,0 +1,115 @@
+/**
+* Copyright 2012-2017, Plotly, Inc.
+* All rights reserved.
+*
+* This source code is licensed under the MIT license found in the
+* LICENSE file in the root directory of this source tree.
+*/
+
+'use strict';
+
+var Lib = require('../../lib');
+var attributes = require('./attributes');
+var parcatConstants = require('./constants');
+var colorbarDefaults = require('../../components/colorbar/defaults');
+
+function markerDefaults(traceIn, traceOut, defaultColor, layout, coerce) {
+
+    coerce('marker.color', defaultColor);
+
+    if(traceIn.marker) {
+        coerce('marker.cmin');
+        coerce('marker.cmax');
+        coerce('marker.cauto');
+        coerce('marker.colorscale');
+        coerce('marker.showscale');
+        coerce('marker.shape');
+        colorbarDefaults(traceIn.marker, traceOut.marker, layout);
+    }
+}
+
+function dimensionsDefaults(traceIn, traceOut) {
+    var dimensionsIn = traceIn.dimensions || [],
+        dimensionsOut = traceOut.dimensions = [];
+
+    var dimensionIn, dimensionOut, i;
+    var commonLength = Infinity;
+
+    if(dimensionsIn.length > parcatConstants.maxDimensionCount) {
+        Lib.log('parcats traces support up to ' + parcatConstants.maxDimensionCount + ' dimensions at the moment');
+        dimensionsIn.splice(parcatConstants.maxDimensionCount);
+    }
+
+    function coerce(attr, dflt) {
+        return Lib.coerce(dimensionIn, dimensionOut, attributes.dimensions, attr, dflt);
+    }
+
+    for(i = 0; i < dimensionsIn.length; i++) {
+        dimensionIn = dimensionsIn[i];
+        dimensionOut = {};
+
+        if(!Lib.isPlainObject(dimensionIn)) {
+            continue;
+        }
```

Comment:
Templates https://github.com/plotly/plotly.js/pull/2761 imposed a few new constraints on container arrays - in particular the need for a `visible` attribute instead of skipping faulty entries, but it DRY's up this pattern too. See for example the changes to [`parcoords/defaults`](https://github.com/plotly/plotly.js/pull/2761/files#diff-ab474a3235cfcf89be148243bb44c1d1) there.

Reply:
Done in https://github.com/jonmmease/plotly.js/pull/1/commits/2c14168928ccf500dafa3f44ed555944b3c20374 and https://github.com/jonmmease/plotly.js/pull/1/commits/c8e3cc9456a628e8b58f40571943e1417609602a